torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -7164,4 +7165,549 @@ Tensor values_backward(const Tensor& grad, const Tensor& self) {
+  // Even partial numerator used in the continued fraction for betainc.
+  /*
+   * This function computes the partial numerator d_{2m} that is specified
+   * here: https://dlmf.nist.gov/8.17.E23
+   */
+  auto options = at::TensorOptions().dtype(x.dtype()).device(x.device());
+  const Tensor one = at::scalar_tensor(1.0, options);
+  const Tensor two = at::scalar_tensor(2.0, options);
+  int32_t m = iteration;
+  Tensor a_plus_2m = a + two * m;
+  Tensor a_plus_2m_minus_one = a_plus_2m - one;
+  Tensor denominator = a_plus_2m * a_plus_2m_minus_one;
+
+  Tensor db = m * x / denominator;
+  Tensor value = db * (b - m);
+  Tensor da = -value * (a_plus_2m + a_plus_2m_minus_one) / denominator;
+
+  return std::make_tuple(std::move(value), at::cat({std::move(da), std::move(db)}, -1));
+}
+
+std::tuple<Tensor, Tensor> _betainc_odd_partial_numerator(const int32_t iteration, const Tensor& a, const Tensor& b, const Tensor& x) {
+  // Odd partial numerator used in the continued fraction for betainc.
+  /*
+   * This function computes the partial numerator d_{2m + 1} that is specified
+   * here: https://dlmf.nist.gov/8.17.E23
+   */
+  auto options = at::TensorOptions().dtype(x.dtype()).device(x.device());
+  const Tensor one = at::scalar_tensor(1.0, options);
+  const Tensor two = at::scalar_tensor(2.0, options);
+  int32_t m = iteration;
+  Tensor a_plus_m = a + m;
+  Tensor a_plus_2m = a_plus_m + m;
+  Tensor a_plus_2m_plus_one = a_plus_2m + one;
+  Tensor a_plus_b_plus_m = a_plus_m + b;
+  Tensor denominator = a_plus_2m * a_plus_2m_plus_one;
+
+  Tensor db = -a_plus_m * x / denominator;
+  Tensor value = db * a_plus_b_plus_m;
+  Tensor da = -value * ((a_plus_2m + a_plus_2m_plus_one) / denominator) - x * (two * a_plus_m + b) / denominator;
+
+  return std::make_tuple(std::move(value), at::cat({std::move(da), std::move(db)}, -1));
+}
+
+std::tuple<Tensor, Tensor, Tensor> _betainc_modified_lentz_method(const Tensor& a, const Tensor& b, const Tensor& x, const Tensor& use_continued_fraction) {
+  // Returns the continued fraction for betainc by modified Lentz's method.
+  /*
+   * This function implements the method described in the appendix of [1] for
+   * evaluating continued fractions.
+   * [1] Thompson, Ian J., and A. Ross Barnett.
+   *     Coulomb and Bessel functions of complex arguments and order.
+   *     Journal of Computational Physics 64.2 (1986): 490-509.
+   *     https://www.fresco.org.uk/papers/Thompson-JCP64p490.pdf
+   */
+  // a, b, and x have same dtype
+  std::tuple<Tensor, Tensor> eps_tiny = AT_DISPATCH_FLOATING_TYPES_AND2(
+        at::ScalarType::Half,
+        at::ScalarType::BFloat16,
+        x.scalar_type(),
+        "__betainc_modified_lentz_method_eps_tiny",
+        [&]() -> std::tuple<Tensor, Tensor> {
+    Tensor eps = at::scalar_to_tensor(std::numeric_limits<at::scalar_value_type<scalar_t>::type>::epsilon(), x.device());
```

Comment:
Not sure why you use the AT_DISPATCH_FLOATING_TYPES_AND2 macro here. You can get these values from x.scalar_type() directly no?

Reply:
I referenced the below code snippet in **torch/csrc/TypeInfo.cpp** for getting eps, tiny, max, min, etc of floating type.
```
#define _AT_DISPATCH_FINFO_TYPES(TYPE, NAME, ...) \
  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND6(    \
      at::kHalf,                                  \
      at::ScalarType::BFloat16,                   \
      at::ScalarType::Float8_e5m2,                \
      at::ScalarType::Float8_e5m2fnuz,            \
      at::ScalarType::Float8_e4m3fn,              \
      at::ScalarType::Float8_e4m3fnuz,            \
      TYPE,                                       \
      NAME,                                       \
      __VA_ARGS__)

static PyObject* THPFInfo_eps(THPFInfo* self, void*) {
  HANDLE_TH_ERRORS
  return _AT_DISPATCH_FINFO_TYPES(self->type, "epsilon", [] {
    return PyFloat_FromDouble(
        std::numeric_limits<at::scalar_value_type<scalar_t>::type>::epsilon());
  });
  END_HANDLE_TH_ERRORS
}
```
The scalar type seems to be given to the template  `std::numeric_limits<at::scalar_value_type<scalar_t>::type>::epsilon()` in compile time, to get epsilon value.

I don't know how to get floating numbers' information(eps, tiny, min, max, etc)  in run time.
If you let me know how to get the float information in run time, I'll use it.

Reply:
> On GPU it might be more performant to always do the 100 iterations here? Maybe we could do more fancy async checking of these conditions, but I would definitely keep that for a follow up PR.
For CPU, this will be good for sure.

I think it might be not good to remove control flow with maximum iteration loop.
When I trained model(LBD, recsys23), the train iteration becomes slower and slower.(2~4 times?, not accurate)
In tensorflow-probability, same situation occurs (slower iterations), which means bugs in PyTorch didn't make iterations slower, I think.
I guess the break point becomes longer in the indefinite loop in betainc's bwd as training process is going on in LBD(recsys23) model.


So, I think it's better to do more fancy async checking that you said than to execute maximum iterations statically(loop unrolling?). I think the fancy async checking can be done in follow-up PR, as you said.